File: appinventor/components-ios/src/LayoutType.swift
// -*- mode: java; c-basic-offset: 2; -*-
// Copyright © 2025 MIT, All rights reserved
// Released under the Apache License, Version 2.0
// https://www.apache.org/licenses/LICENSE-2.0

@objc public class LayoutType: NSObject, OptionList {
  @objc public static let MainText = LayoutType(0)
  @objc public static let MainText_DetailText_Vertical = LayoutType(1)
  @objc public static let MainText_DetailText_Horizontal = LayoutType(2)
  @objc public static let Image_MainText = LayoutType(3)
  @objc public static let Image_MainText_DetailText_Vertical = LayoutType(4)
  @objc public static let ImageTop_MainText_DetailText = LayoutType(5)

  private static let LOOKUP: [Int32: LayoutType] = generateOptionsLookup(MainText, MainText_DetailText_Vertical, MainText_DetailText_Horizontal, Image_MainText, Image_MainText_DetailText_Vertical, ImageTop_MainText_DetailText)

  let value: Int32

  @objc private init(_ value: Int32) {
    self.value = value
  }

  @objc public func toUnderlyingValue() -> AnyObject {
    return value as AnyObject
  }

  @objc public static func fromUnderlyingValue(_ value: Int32) -> LayoutType? {
    return LOOKUP[value]
  }
}

File: appinventor/components/src/com/google/appinventor/components/common/LayoutType.java
// -*- mode: java; c-basic-offset: 2; -*-
// Copyright 2023 MIT, All rights reserved
// Released under the Apache License, Version 2.0
// http://www.apache.org/licenses/LICENSE-2.0

package com.google.appinventor.components.common;

import java.util.HashMap;
import java.util.Map;

/**
 * Defines a parameter of ListViewLayout used by the ListView component.
 */
public enum LayoutType implements OptionList<Integer> {
  @Default
  MainText(0),
  MainText_DetailText_Vertical(1),
  MainText_DetailText_Horizontal(2),
  Image_MainText(3),
  Image_MainText_DetailText_Vertical(4),
  ImageTop_MainText_DetailText(5);

  private final int layout;

  LayoutType(int value) {
    this.layout = value;
  }

  public Integer toUnderlyingValue() {
    return layout;
  }

  private static final Map<Integer, LayoutType> lookup = new HashMap<>();

  static {
    for (LayoutType value : LayoutType.values()) {
      lookup.put(value.toUnderlyingValue(), value);
    }
  }

  public static LayoutType fromUnderlyingValue(Integer value) {
    return lookup.get(value);
  }
}
